### PR TITLE
ci: Detect breaking changes in packages other than Uno.[UI|WinUI]

### LINKED ISSUE #11548: Breaking changes in `Uno.UI.Runtime.XYZ` projects are not recognized
### Current behavior

When a breaking change occurs in projects like `Uno.UI.Runtime.Skia.Wpf`, it is not marked in the CI.

E.g. - https://github.com/unoplatform/uno/pull/10935/files#diff-e79067b16894fd0c02a5153595b91726b9ea0570810ecd42f8b9d09a94941e67R6-R11

### Expected behavior

Diffignore update should be required

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

None

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #11548, closes https://github.com/unoplatform/uno/issues/10928

## PR Type

What kind of change does this PR introduce?

- Build or CI related changes

## What is the current behavior?

Breaking changes are not detected for packages other than Uno.UI and Uno.WinUI.

## What is the new behavior?

They are detected now.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.Runtime.Skia.Wpf/Controls/WpfTextViewTextBox.cs [hunk 1]
```diff
@@ -3,12 +3,12 @@
 
 namespace Uno.UI.Runtime.Skia.Wpf.Controls;
 
-internal class WpfTextViewTextBox : TextBox
+internal class WpfTextViewTextBox2 : TextBox
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Controls/WpfTextViewTextBox.cs:6]: Oh this **was** public in past. I will find another public type to test the change on.

**Comment by Youssef1313** [comment]
It's working!

```
  Comparing C:\Users\AzDevOps\AppData\Local\Temp\8b6e03e7-2268-4bf8-9adc-7b1e437b3229\lib\net47\Uno.UI.Runtime.Skia.Wpf.dll and C:\Users\AzDevOps\AppData\Local\Temp\6fe28c55-9ad6-4394-a5f3-0effcbd437c5\lib\net47\Uno.UI.Runtime.Skia.Wpf.dll
##[error]EXEC(0,0): Error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set.
EXEC : error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  Comparing C:\Users\AzDevOps\AppData\Local\Temp\8b6e03e7-2268-4bf8-9adc-7b1e437b3229\lib\net7.0-windows7.0\Uno.UI.Runtime.Skia.Wpf.dll and C:\Users\AzDevOps\AppData\Local\Temp\6fe28c55-9ad6-4394-a5f3-0effcbd437c5\lib\net7.0-windows7.0\Uno.UI.Runtime.Skia.Wpf.dll
##[error]EXEC(0,0): Error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set.
EXEC : error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  Comparing C:\Users\AzDevOps\AppData\Local\Temp\8b6e03e7-2268-4bf8-9adc-7b1e437b3229\lib\netcoreapp3.1\Uno.UI.Runtime.Skia.Wpf.dll and C:\Users\AzDevOps\AppData\Local\Temp\6fe28c55-9ad6-4394-a5f3-0effcbd437c5\lib\netcoreapp3.1\Uno.UI.Runtime.Skia.Wpf.dll
##[error]EXEC(0,0): Error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set.
EXEC : error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  Done comparing, with differences.
##[error]EXEC(0,0): Error : Build failed with unexpected differences. Modifications to the public API introduce binary breaking changes and should be avoided.
EXEC : error : Build failed with unexpected differences. Modifications to the public API introduce binary breaking changes and should be avoided. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  If these modifications were expected and intended, see https://github.com/nventive/Uno.PackageDiff#how-to-provide-an-ignore-set on how to ignore them.
##[error]Build\Uno.UI.Build.csproj(499,3): Error MSB3073: The command "C:\a\1\s\Build\\tools\generatepkgdiff.exe --base=Uno.UI.Runtime.Skia.Wpf --other=..\..\..\NugetPackages-Artifacts-skia-UWP\vslatest-skia\Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.nupkg --diffignore=PackageDiffIgnore.xml --outfile=C:\a\1\a\ApiDiff.Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.md" exited with code 1.
C:\a\1\s\Build\Uno.UI.Build.csproj(499,3): error MSB3073: The command "C:\a\1\s\Build\\tools\generatepkgdiff.exe --base=Uno.UI.Runtime.Skia.Wpf --other=..\..\..\NugetPackages-Artifacts-skia-UWP\vslatest-skia\Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.nupkg --diffignore=PackageDiffIgnore.xml --outfile=C:\a\1\a\ApiDiff.Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.md" exited with code 1.
##[error]Process 'msbuild.exe' exited with code '1'.
```

File: src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs [hunk 2]
```diff
@@ -7,15 +7,15 @@
 
 namespace Uno.UI.Runtime.Skia.Wpf
 {
-	public class WpfApplicationExtension : IApplicationExtension, IDisposable
+	public class WpfApplicationExtension2 : IApplicationExtension, IDisposable
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs:10]:
```suggestion
	public class WpfApplicationExtension: IApplicationExtension, IDisposable
```

File: src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs [hunk 3]
```diff
@@ -7,15 +7,15 @@
...
 	{
 		private readonly Application _owner;
 		private WpfSystemThemeHelperExtension _themeHelper;
 		private SystemTheme _currentTheme;
 
 		private bool _disposedValue;
 
-		public WpfApplicationExtension(Application owner)
+		public WpfApplicationExtension2(Application owner)
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs:18]:
```suggestion
		public WpfApplicationExtension(Application owner)
```

File: src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs [hunk 4]
```diff
@@ -54,7 +54,7 @@ protected virtual void Dispose(bool disposing)
 			}
 		}
 
-		~WpfApplicationExtension()
+		~WpfApplicationExtension2()
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs:57]:
```suggestion
		~WpfApplicationExtension()
```

File: src/Uno.UI.Runtime.Skia.Wpf/WpfHost.cs [hunk 5]
```diff
@@ -84,7 +84,7 @@ internal static void RegisterExtensions()
 
 			ApiExtensibility.Register(typeof(Uno.ApplicationModel.Core.ICoreApplicationExtension), o => new CoreApplicationExtension(o));
 			ApiExtensibility.Register(typeof(Windows.UI.Core.ICoreWindowExtension), o => new WpfCoreWindowExtension(o));
-			ApiExtensibility.Register<Windows.UI.Xaml.Application>(typeof(IApplicationExtension), o => new WpfApplicationExtension(o));
+			ApiExtensibility.Register<Windows.UI.Xaml.Application>(typeof(IApplicationExtension), o => new WpfApplicationExtension2(o));
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/WpfHost.cs:87]:
```suggestion
			ApiExtensibility.Register<Windows.UI.Xaml.Application>(typeof(IApplicationExtension), o => new WpfApplicationExtension(o));
```

File: build/Uno.UI.Build.csproj [hunk 6]
```diff
@@ -482,7 +482,28 @@
...
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix).Lottie --other=$(PackageNamePrefix).Lottie.$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).Lottie.$(GITVERSION_SemVer).md" />
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix).Maps --other=..\..\..\NugetPackages-Artifacts-generic-UWP\vslatest-generic\$(PackageNamePrefix).Maps.$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).Maps.$(GITVERSION_SemVer).md" />
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix).MSAL --other=$(PackageNamePrefix).MSAL.$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).MSAL.$(GITVERSION_SemVer).md" />
```
Comment (jeromelaban) [inline build/Uno.UI.Build.csproj:489]: We should be able to use something like that:

```suggestion
        <ItemGroup>
           <_diffPackage Include="$(PackageNamePrefix).$(GITVERSION_SemVer)" />
            <_diffPackage Include="$(PackageNamePrefix).Foldable.$(GITVERSION_SemVer)" />
        </ItemGroup>

		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix) --other=%(_diffPackage .Identity).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.%(_diffPackage .Identity).md" />
```

**Review by jeromelaban — APPROVED** [review]
🚀

File: build/Uno.UI.Build.csproj [hunk 7]
```diff
@@ -481,8 +481,33 @@
...
 		</PropertyGroup>
 
+        <ItemGroup>
+			<_diffPackage Include="$(PackageNamePrefix)" Other="$(PackageNamePrefix).$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).Foldable" Other="..\..\..\NugetPackages-Artifacts-generic-UWP\vslatest-generic\$(PackageNamePrefix).Foldable.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).Lottie" Other="$(PackageNamePrefix).Lottie.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).Maps" Other="..\..\..\NugetPackages-Artifacts-generic-UWP\vslatest-generic\$(PackageNamePrefix).Maps.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).MSAL" Other="$(PackageNamePrefix).MSAL.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).Svg" Other="$(PackageNamePrefix).Svg.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).RemoteControl" Other="$(PackageNamePrefix).RemoteControl.$(GITVERSION_SemVer).nupkg" />
```
Comment (Youssef1313) [inline build/Uno.UI.Build.csproj:491]: I assume I should remove RemoteControl as its APIs are not directly used by customers and we should freely change it? @jeromelaban Can you confirm please?

Comment (Youssef1313) [reply build/Uno.UI.Build.csproj]: Removed for now. We can add if necessary later.

**Comment by Youssef1313** [comment]
This is failing with:

```
EXEC(0,0): Error : Removed method System.Uri Microsoft.Toolkit.Uwp.UI.Lottie.LottieVisualSourceBase.get_UriSource() not found in ignore set.
EXEC(0,0): Error : Removed method System.Void Microsoft.Toolkit.Uwp.UI.Lottie.LottieVisualSourceBase.set_UriSource(System.Uri value) not found in ignore set.
```

I don't really understand, UriSource doesn't seem to be removed https://github.com/unoplatform/uno/blob/bdd49d9b5ec46c205c67b4903bfa9e5d99702627/src/AddIns/Uno.UI.Lottie/LottieVisualSourceBase.cs#L42-L46

@jeromelaban @Robert-Louis Any idea?

**Comment by jeromelaban** [comment]
> @jeromelaban @Robert-Louis Any idea?

This is likely caused by a change in the implementation. `UriSource` is now a member of an interface (where it previously was not), and it's likely changing its actual definition.

**Comment by Robert-Louis** [comment]
> This is likely caused by a change in the implementation. `UriSource` is now a member of an interface (where it previously was not), and it's likely changing its actual definition.

@jeromelaban how should we fix it ? Should I find a way to not adding it to the interface ? 
I not sure how the CI work with this.

**Comment by Youssef1313** [comment]
@jeromelaban That was my first guess, but looking into IL for a "simplified" case, I see that the call in both these cases produce the same IL

image

Maybe I'm missing something?

**Pull request merged**